### ISSUE TITLE
Documentation Content: Versioning

### ISSUE THREAD

**Issue by nate-double-u** [etcd-io/website#82]
As per the Documentation Content Section of the [Etcd.io Docs/SEO Improvement Plan](https://github.com/etcd-io/website/issues/65)

- [ ] Fix multi-version navigation
  - That is, switching "versions" of the documentation should automatically navigate to that page in the selected version of the docs, if that page exists. 
  - If this is challenging to do with the documentation toolchain we have, re-evaluate whether we want to keep versioned documentation.

**Comment by nate-double-u** [comment]
I've seen a couple versioning systems around Hugo in the CNCF ecosystem.

Subdomain versioning
* [Kubernetes website](https://kubernetes.io) 
* [CNI](https://www.cni.dev) - @nate-double-u put this one together, and wrote up a [versioning howto](https://github.com/containernetworking/cni.dev#versioning-the-cni-website). The Kubernetes system is similar.

Folder versioning
* [Thanos](https://thanos.io/) - seems to have a version selector that works the way we want -- Interestingly, it looks like their website is fully contained in their [code repo](https://github.com/thanos-io/thanos/tree/master/website).
* [Longhorn](https://longhorn.io) - seems not to have a version selector that works the way we want

There are also several sites that don't version (I won't list them here).

Kubernetes uses a versioning system that requires a subdomain ([as does Docsy ](https://www.docsy.dev/docs/adding-content/versioning/)), which makes releases a bit more work, but it does seem to solve the switching issue mentioned above. I'm pretty sure the switching issue can be sorted either way -- but it seems like the Docsy/Kubernetes route gets us there easiest (though, there would be other work to do to port the site, so there's no free versioning that I'm aware of).

@jberkus @spzala

**Comment by nate-double-u** [comment]
@sftim I know you're not on this project, but we have talked about Hugo versioning before. You mentioned that you've got a point of view on it, and I was wondering what your opinion is regarding best practices for versioning a Hugo site.

**Comment by nate-double-u** [comment]
@chalin You've just finished helping gRPC do a [Docsy migration](https://github.com/grpc/grpc.io/issues/479). I don't see any versioning on the site, are there plans for versioning their docs?

**Comment by sftim** [comment]
> Folder versioning

I think usually I'd prefer to go with the end-user experience that Thanos is offering, or something close to it. [`websitepreprocess.sh`](https://github.com/thanos-io/thanos/blob/6a183d0720393b8100882d115fb566afd9774470/scripts/website/websitepreprocess.sh) seems to be where the magic happens - credit to GitHub user @thisisobate.

What doesn't fit from the Longhorn approach? I'm guessing it's about letting you switch to historical versions of the same page?

**Comment by nate-double-u** [comment]
> What doesn't fit from the Longhorn approach? I'm guessing it's about letting you switch to historical versions of the same page?

That exactly, and I think that there is a reasonably easy fix for that (I should open an issue). I ran into the same sort of problem when building out the CNI site.

> [websitepreprocess.sh](https://github.com/thanos-io/thanos/blob/6a183d0720393b8100882d115fb566afd9774470/scripts/website/websitepreprocess.sh) seems to be where the magic happens

I'll definitely be taking a look at that

**Comment by nate-double-u** [comment]
@jberkus you'd mentioned that potentially there's a third way for versioning, where we wouldn't use the version in the URL at all. Is this better for SEO? I haven't seen any other CNCF sites doing it this way, but I'd be happy to research it.

**Comment by nate-double-u** [comment]
Looking more closely at the Thanos.io model -- it looks like it also brings visitors back to the getting started page on version selection.

**Comment by nate-double-u** [comment]
I've also been looking at cookies or potentially using query strings for managing versioning.

I think that using cookies is problematic as they'd break bookmarking.

Using query strings wouldn't break bookmarking, but I need to do some more research into how to implement them -- I don't think Hugo would manage them, but Netlify maybe able to help there. I'll report back when I know more.

**Comment by sftim** [comment]
Static site generators and versions in query strings don't seem like they are going to play well with each other. There are some places where putting a version in a query string though.

**Comment by nate-double-u** [comment]
TL;DR: Query strings don't solve our problem. Still research to be done.

I've done some testing with Netlify's redirects, and I agree @sftim, this is likely not the way to go.

You can see some of the work here: https://github.com/nate-double-u/cncf-hugo-netlify-starter/tree/versions-sandbox

Because these are just redirects, there's not a whole lot of smarts, and while I think I could probably write a function that the redirects could use I'm not sure it's worth it. The redirects here work going one way, but if i try to set them up to work the other way I get an infinite loop :/

https://github.com/nate-double-u/cncf-hugo-netlify-starter/blob/762a525ce4bc9c248e5938becb985b3abc9c8abd/netlify.toml#L14-L26

main index redirect: https://stoic-euler-268859.netlify.app/docs/test/?version=v0.0.1
a test page further in the path: https://stoic-euler-268859.netlify.app/docs/test/?version=v0.0.1&path=test

I don't see this as being any less work than just coding the selector snippet to do this for us. I've done this with the https://www.cni.dev site, but because it was using urls for the versioning, these snippets work:

https://github.com/containernetworking/cni.dev/blob/5373bbe7a72567c30a1104602a2aad9065b3b75d/layouts/partials/deprecation-warning.html#L10

https://github.com/containernetworking/cni.dev/blob/5373bbe7a72567c30a1104602a2aad9065b3b75d/layouts/partials/navbar-version-selector.html#L8-L14

Here it's easy to pick out the version because the info about the version is a part of the `site.Params` list (see https://github.com/containernetworking/cni.dev/blob/5373bbe7a72567c30a1104602a2aad9065b3b75d/config.toml#L34).

What I'm now not sure of is how to parse the `.RelPermalink` as the version is going to be embedded in it, rather than being it's own piece that can be referenced.

(And i'm again looking for an example of a folder-style versioning system that handles historical versions of the same page correctly).

**Comment by nate-double-u** [comment]
Looks like Velero.io is doing their versioning the way we'd like ours to work: https://velero.io/docs/v1.5/
(and they've been kind enough to show us how they did it: https://github.com/vmware-tanzu/velero/tree/main/site/layouts/docs)

**Comment by nate-double-u** [comment]
I've tried to build nav fix, but there are a couple todo items that I'm not sure that a. we need to handle, and b. how we should handle them if we do.

https://github.com/etcd-io/website/pull/86#issuecomment-752310501
(Mostly around the structural differences between the v2 and v3 documentation and how to deal with the resulting 404s)

Tangent: If we want to migrate the documentation from the code repo to the website repo, we should probably open an issue, or make an addendum to the [Etcd.io Docs/SEO Improvement Plan](https://github.com/etcd-io/website/issues/65).

**Comment by thisisobate** [comment]
> > Folder versioning
> 
> I think usually I'd prefer to go with the end-user experience that Thanos is offering, or something close to it. [`websitepreprocess.sh`](https://github.com/thanos-io/thanos/blob/6a183d0720393b8100882d115fb566afd9774470/scripts/website/websitepreprocess.sh) seems to be where the magic happens - credit to GitHub user @thisisobate.
> 
> What doesn't fit from the Longhorn approach? I'm guessing it's about letting you switch to historical versions of the same page?

Hey @sftim @nate-double-u
Glad to see that my work on docs versioning at Thanos is being considered as a potential solution here.

I would also like to point out that we are continuing the work on docs versioning by making it a stand-alone tool which we called [Mdox](https://github.com/bwplotka/mdox)- please feel free to check it out.

We hope to make the tool more production-ready and stable. 

To better understand how we came about with our "folder versioning" approach, please feel free to check out the proposal/design doc [here](https://thanos.io/tip/proposals/202005_version_documentation.md/)

**Comment by sftim** [comment]
Hugo has [rendering hooks](https://gohugo.io/getting-started/configuration-markup/) for templates including `render-link`. Might those help?

**Comment by nate-double-u** [comment]
In the [second PR I posted](https://github.com/etcd-io/website/pull/86/files#diff-0548eb28dd1d566de5085ec0194b594abc38c131bf64704d35e4445ab75221cd), I used a `replace` fn call after figuring out what the current version and new versions should be for any given link.

The downside to this is having to add a version to each top level `_index.md` file, but this isn't as bad as it could have been -- the `cascade` allows this to propagate throughout the version's folder. We'll likely need to go back and add this to the `Documentation` folder in the code repo at some point.

Frontmatter version example
```
---
cascade:
  version: v3.3.13
---
```


> Hugo has [rendering hooks](https://gohugo.io/getting-started/configuration-markup/) for templates including `render-link`. Might those help?

I'll definitely take a look and see if they can help make the versioning even neater.

**Comment by nate-double-u** [comment]
Latest push removes the "v" from the urls in `layouts/index.redirects`. This fixes the 404 page's `/docs/latest` link.
